core/src/main/java/com/linecorp/armeria/common/logging/RequestScopedMdc.java

```diff
@@ -0,0 +1,363 @@
+ * before the thread-local context map:
+ * <pre>{@code
+ * RequestContext ctx = ...;
+ * RequestScopedMdc.put(ctx, "transactionId", "1234");
+ * try (SafeCloseable ignored = ctx.push()) {
+ *     assert MDC.get("transactionId").equals("1234");
+ *
+ *     // A request-scoped property always gets higher priority:
+ *     MDC.put("transactionId", "5678");
+ *     assert MDC.get("transactionId").equals("1234");
+ * }
+ *
+ * // Now using the thread-local property
+ * // because not in a request scope anymore
+ * assert MDC.get("transactionId").equals("5678");
+ * }</pre></p>
+ */
+public final class RequestScopedMdc {
+
+    private static final Logger logger = LoggerFactory.getLogger(RequestScopedMdc.class);
+
+    private static final AttributeKey<Map<String, String>> MAP =
+            AttributeKey.valueOf(RequestScopedMdc.class, "map");
+
+    private static final String ERROR_MESSAGE =
+            "Failed to replace the " + MDCAdapter.class.getSimpleName() + "; " +
+            RequestScopedMdc.class.getSimpleName() + " will not work.";
+
+    @Nullable
+    private static final MDCAdapter delegate;
+
+    static {
+        // Trigger the initialization of the default MDC adapter.
+        MDC.get("");
+
+        // Replace the default MDC adapter with ours.
+        MDCAdapter oldAdapter = null;
+        try {
+            final Field mdcAdapterField = MDC.class.getDeclaredField("mdcAdapter");
+            mdcAdapterField.setAccessible(true);
+            oldAdapter = (MDCAdapter) mdcAdapterField.get(null);
+            mdcAdapterField.set(null, new Adapter(oldAdapter));
+        } catch (Throwable t) {
+            logger.warn(ERROR_MESSAGE, t);
+        }
+        delegate = oldAdapter;
+    }
+
+    /**
+     * Returns the value of the specified request-scoped {@link MDC} property bound to the specified
+     * {@link RequestContext}..
+     *
+     * @param ctx the {@link RequestContext}
+     * @param key the key of the request-scoped {@link MDC} property
+     *
+     * @return the request-scoped {@link MDC} property. {@code null} if not found.
+     */
+    @Nullable
+    public static String get(RequestContext ctx, String key) {
+        requireNonNull(ctx, "ctx");
+        requireNonNull(key, "key");
+        return getMap(ctx).get(key);
+    }
+
+    /**
+     * Returns the {@link Map} of all request-scoped {@link MDC} properties bound to the specified
+     * {@link RequestContext}..
+     *
+     * @param ctx the {@link RequestContext}
+     *
+     * @return the {@link Map} that contains all request-scoped {@link MDC} properties.
+     *         An empty {@link Map} if there are no request-scoped {@link MDC} properties.
+     */
+    public static Map<String, String> getAll(RequestContext ctx) {
+        requireNonNull(ctx, "ctx");
+        final Map<String, String> map = getMap(ctx);
+        // Note: We ensure an empty map is always immutable.
+        return map.isEmpty() ? map : Collections.unmodifiableMap(map);
```

Comment:
Isn't it better to always put `unmodifiableMap` to the context and return the map without making it unmodifiable? because the implementation supposes that reading is a lot than writing?